### PR TITLE
Add delay instruction

### LINKED ISSUE #2846: Add a blocking Delay instruction to pulse
### What is the expected enhancement?
We should add a new type of instruction `Delay` to pulse. This will be a blocking instruction that will occupy the timeslot for a given duration on the specified channel. This command should accept only duration and name as parameters. The command will be ignored at assembly time.

### PR DESCRIPTION
**Pull request by taalexander**
### Summary
Closes #2846. This PR adds a delay instruction to the pulse module. 

### Details and comments
Delay's are used to block out times in a pulse schedule. This is useful for cases such as scheduling circuits to pulses with barriers.

Delay's are not transmitted to the backend since instructions within a Qobj are scheduled absolutely in time.

### CODE REVIEW

File: qiskit/pulse/commands/delay.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+class Delay(Command):
+    """Delay command."""
+
+    prefix = 'delay'
+
+    def __init__(self, duration: int, name: Optional[str] = None):
+        """Create a new delay command.
+
+        No other may be scheduled within a delay command.
```
Comment (lcapelluto) [inline qiskit/pulse/commands/delay.py]:
```suggestion
        No other command may be scheduled within a delay command.
```

File: qiskit/pulse/commands/delay.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+        Delays that exist as the last instruction on a channel will be removed.
+
+        Note: Delay's are not explicitly sent to backends with the current Qobj
```
Comment (lcapelluto) [inline qiskit/pulse/commands/delay.py]:
```suggestion
        Note: Delays are not explicitly sent to backends with the current Qobj
```

File: qiskit/pulse/commands/delay.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        transport layer. They are only used to enforce the correct offset in time
+        for instructions that will be submitted directly to the backend.
+
+        Args:
+            duration: int
```
Comment (lcapelluto) [inline qiskit/pulse/commands/delay.py]:
```suggestion
            duration: Length of time of the delay in terms of dt.
```

File: test/python/pulse/test_schedule.py [hunk 4]
```diff
@@ -290,22 +294,158 @@ def test_name_inherited(self):
...
+
+        drive_ch = self.two_qubit_device.qubits[0].drive
+        pulse = SamplePulse(np.full(10, 0.1))
+        # should pass as is an append
+        sched = self.delay(drive_ch) + pulse(drive_ch)
+        self.assertIsInstance(sched, Schedule)
+        # should fail due to overlap
+        with self.assertRaises(PulseError):
+            sched = self.delay(drive_ch) | pulse(drive_ch)
+            self.assertIsInstance(sched, Schedule)
```
Comment (lcapelluto) [inline test/python/pulse/test_schedule.py]: This line shouldn't be reached, right?

File: test/python/pulse/test_schedule.py [hunk 5]
```diff
@@ -290,22 +294,158 @@ def test_name_inherited(self):
...
+
+        measure_ch = self.two_qubit_device.qubits[0].measure
+        pulse = SamplePulse(np.full(10, 0.1))
+        # should pass as is an append
+        sched = self.delay(measure_ch) + pulse(measure_ch)
+        self.assertIsInstance(sched, Schedule)
+        # should fail due to overlap
+        with self.assertRaises(PulseError):
+            sched = self.delay(measure_ch) | pulse(measure_ch)
+            self.assertIsInstance(sched, Schedule)
```
Comment (lcapelluto) [inline test/python/pulse/test_schedule.py]: same here?

File: test/python/visualization/test_pulse_visualization_output.py [hunk 6]
```diff
@@ -105,7 +107,6 @@ def test_schedule_matplotlib_drawer(self):
         schedule = self.sample_schedule()
         pulse_drawer(schedule, filename=filename)
         self.assertImagesAreEqual(filename, self.schedule_matplotlib_reference)
-        os.remove(filename)
```
Comment (lcapelluto) [inline test/python/visualization/test_pulse_visualization_output.py]: bring back!

File: test/python/pulse/test_schedule.py [hunk 7]
```diff
@@ -290,22 +294,158 @@ def test_name_inherited(self):
...
+        self.delay = Delay(self.delay_time)
+
+    def test_delay_drive_channel(self):
+        """Test Delay on DriveChannel"""
+
+        drive_ch = self.two_qubit_device.qubits[0].drive
+        pulse = SamplePulse(np.full(10, 0.1))
+        # should pass as is an append
+        sched = self.delay(drive_ch) + pulse(drive_ch)
+        self.assertIsInstance(sched, Schedule)
```
Comment (lcapelluto) [inline test/python/pulse/test_schedule.py]: check that pulse starts at t=10

File: test/python/pulse/test_schedule.py (hunk 4 above)
Comment (taalexander) [reply test/python/pulse/test_schedule.py]: yes, will remove.

**Comment by taalexander** [comment]
Should be good to go now.

**Comment by nkanazawa1989** [comment]
This looks good, but I have one question. If `Delay` is used to block insertion of other instruction, it is not expected to have this in `CmdDef`? Is so, we need `QobjToInstructionConverter` (also spec update required), or maybe in future update?

**Comment by taalexander** [comment]
Yes, but this will be a future PR. Similarly delay will be inserted for
buffers.

File: qiskit/assembler/assemble_schedules.py [hunk 8]
```diff
@@ -76,7 +76,9 @@ def assemble_schedules(schedules, qobj_id, qobj_header, run_config):
                     # verify all acquires satisfy meas_map
                     _validate_meas_map(instruction, meas_map)
 
-            qobj_instructions.append(instruction_converter(shift, instruction))
+            converted_instruction = instruction_converter(shift, instruction)
+            if converted_instruction:
```
Comment (lcapelluto) [inline qiskit/assembler/assemble_schedules.py]: There's some implicit logic here: assume `converted_instruction` will only evaluate to `False` if `instruction` is a `Delay`. Can we check the type of `instruction`, and only convert for non delays? Or in some other way make our failure routes or expected behavior more explicit?

**Pull request merged**